### ISSUE TITLE
Unify the signatures of `fit()` and `transform()` methods

### ISSUE THREAD

**Issue by bakhtiary** [jpmml/sklearn2pmml#133]
Hi,

I wanted to chain two lookup tables but I am having problems. 
The first lookup table works on a single column from a dataframe
The second lookup table is applied to the output of the first lut and another column of the dataframe.
The following illustrates what I'm trying to achieve:
``` python
mlut = MultiLookupTransformer(mapping, -1.0)
hour_lut = LookupTransformer(single_mapping, -1)
mapper = DataFrameMapper([
                ('hour', [hour_lut]),
                ('size', None)
            ])
pipeline = PMMLPipeline(
    [
        ("mapper", mapper),
        ("mlut", mlut),
        ("regressor", lr)
    ]
)
pipeline.fit(data, target)
```

calling fit I receive:
```
/venv/lib/python3.6/site-packages/sklearn/externals/joblib/memory.py in __call__(self, *args, **kwargs)
    340 
    341     def __call__(self, *args, **kwargs):
--> 342         return self.func(*args, **kwargs)
    343 
    344     def call_and_shelve(self, *args, **kwargs):
/venv/lib/python3.6/site-packages/sklearn/pipeline.py in _fit_transform_one(transformer, X, y, weight, **fit_params)
    612 def _fit_transform_one(transformer, X, y, weight, **fit_params):
    613     if hasattr(transformer, 'fit_transform'):
--> 614         res = transformer.fit_transform(X, y, **fit_params)
    615     else:
    616         res = transformer.fit(X, y, **fit_params).transform(X)
/venv/lib/python3.6/site-packages/sklearn/base.py in fit_transform(self, X, y, **fit_params)
    463         else:
    464             # fit method of arity 2 (supervised transformation)
--> 465             return self.fit(X, y, **fit_params).transform(X)
    467 

TypeError: fit() takes 2 positional arguments but 3 were given
```
I wanted to know if there is a correct way to do this?

Best,
Amir

**Comment by vruusmann** [comment]
> TypeError: fit() takes 2 positional arguments but 3 were given

The signature of the fit method is `(Multi)LookupTransformer.fit(self, Y)`. This is different from traditional fit method signatures `fit(self, X, y)`.

The conclusion here is that lookup transformers are not designed to be used as standalone pipeline steps. They should be used inside `DataFrameMapper`, where it is guaranteed that the fit method fill be called using one vector/matrix argument only (pipeline steps are most of the time called using two vector/matrix arguments - one for the label, and another one for features).

The solution is to refactor the first step of your pipeline. It's possible to have multiple `DataFrameMapper` instances there, and their results can be merged using the `FeatureUnion` meta-transformation:
```python
pipeline = PMMLPipeline([
  ("features", FeatureUnion([
    ("first_mapper", DataFrameMapper([ .. ])),
    ("second_mapper", DataFrameMapper([ .. ]))
  ])),
  ("estimator", ..)
])
```

Also, the latest SkLearn2PMML release version 0.42.0 adds support for the `sklearn.compose.ColumnTransformer`:
https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html

This transformer is a bit more flexible than `DataFrameMapper`, and it's possible to chain multiple `ColumnTransformer` steps together without using `FeatureUnion`.

TLDR: Your problem is easily solvable, either by using `FeatureUnion`, or by switching from `DataFrameMapper` (uses indexing by name) to `ColumnTransformer` (uses indexing by name and/or by position). However, you can't insert a `(Multi)LookupTransformer` step directly into pipeline, it needs to have some sort of wrapper around it.

**Comment by vruusmann** [comment]
Reopening, because I can't immediately think of a very good reason why it shouldn't be possible to use `(Multi)LookupTransformer` as a standalone pipeline step.

This decision needs to be documented (in order to avoid confusing my future self again), and applied to all custom transformer classes.
